hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateNftsHandler.java

```diff
@@ -81,8 +89,25 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         final var tokenStore = context.createStore(ReadableTokenStore.class);
         final var token = tokenStore.get(op.tokenOrElse(TokenID.DEFAULT));
         if (token == null) throw new PreCheckException(INVALID_TOKEN_ID);
-        if (token.hasMetadataKey()) {
-            context.requireKey(token.metadataKeyOrThrow());
+
+        final var nftStore = context.createStore(ReadableNftStore.class);
+        if (serialNumbersInTreasury(
+                token.treasuryAccountIdOrThrow(), op.serialNumbers(), nftStore, token.tokenIdOrThrow())) {
+            if (token.hasMetadataKey() && token.hasSupplyKey()) {
+                context.requireKey(oneOf(token.metadataKeyOrThrow(), token.supplyKeyOrThrow()));
+            } else if (token.hasMetadataKey()) {
+                context.requireKey(token.metadataKeyOrThrow());
+            } else if (token.hasSupplyKey()) {
+                context.requireKey(token.supplyKeyOrThrow());
+            } else {
+                throw new PreCheckException(TOKEN_IS_IMMUTABLE);
+            }
+        } else {
+            if (token.hasMetadataKey()) {
+                context.requireKey(token.metadataKeyOrThrow());
+            } else {
+                throw new PreCheckException(TOKEN_HAS_NO_METADATA_KEY);
+            }
```

Comment:
```suggestion
            validateTrue(token.hasMetadataKey(), TOKEN_HAS_NO_METADATA_KEY);
            context.requireKey(token.metadataKeyOrThrow())
```
Any reason not to do this?

Reply:
Agreed can do this. But use `validateTruePrecheck`. 

Similar we can do `validateTruePrecheck(token.hasMetadataKey() 
|| token.hasSupplyKey()....` in the if case also

Reply:
Updated code to make use of this.